s3auth-relay/src/main/java/com/s3auth/relay/BaseFacade.java

```diff
@@ -0,0 +1,435 @@
+import lombok.EqualsAndHashCode;
+import lombok.ToString;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.net.ftp.FTPReply;
+
+/**
+ * Convenient base superclass that creates a request dispatcher object for port
+ * listener implementations such as {@link HttpFacade} or {@link FtpFacade}.
+ *
+ * @author Simon Njenga (simtuje@gmail.com)
+ * @author Yegor Bugayenko (yegor@tpc2.com)
+ * @version $Id$
+ * @since 0.1
+ */
+@ToString
+@EqualsAndHashCode(of = { "sockets", "server", "secured" })
+@SuppressWarnings({ "PMD.DoNotUseThreads", "PMD.TooManyMethods" })
+@Loggable(Loggable.DEBUG)
+class BaseFacade implements Closeable {
+
+    /**
+     * How many threads to use.
+     */
+    private static final int THREADS = Tv.HUNDRED;
+
+    /**
+     * Executor service, with socket openers.
+     */
+    private final transient ScheduledExecutorService frontend;
+
+    /**
+     * Executor service, with consuming threads.
+     */
+    private final transient ScheduledExecutorService backend;
+
+    /**
+     * Blocking queue of ready-to-be-processed sockets.
+     */
+    private final transient BlockingQueue<Socket> sockets =
+        new SynchronousQueue<Socket>();
+
+    /**
+     * Server socket.
+     */
+    private final transient ServerSocket server;
+
+    /**
+     * Secured Server socket.
+     */
+    private final transient ServerSocket secured;
+
+    /**
+     * Protected constructor.
+     * @param frontthreads The number of front threads
+     * @param frontprefix The name of the front threads
+     * @param backthreads The number of back threads
+     * @param backprefix The name of the back threads
+     * @param svr The server socket
+     * @param secure The secure server socket
+     * @checkstyle ParameterNumber (5 lines)
+     */
+    protected BaseFacade(final int frontthreads, final String frontprefix,
+        final int backthreads, final String backprefix, final ServerSocket svr,
+        final ServerSocket secure) {
+        this.frontend = this.createThreadPool(frontthreads, frontprefix);
+        this.backend = this.createThreadPool(backthreads, backprefix);
+        this.server = svr;
+        this.secured = secure;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void close() throws IOException {
+        try {
+            this.shutdown(this.frontend);
+            this.shutdown(this.backend);
+        } catch (final InterruptedException ex) {
+            Thread.currentThread().interrupt();
+            throw new IOException(ex);
+        } finally {
+            IOUtils.closeQuietly(this.server);
+            IOUtils.closeQuietly(this.secured);
+        }
+    }
+
+    /**
+     * Start listening to the ports.
+     *
+     * The default implementation is <em>empty</em>. This should be overridden
+     * in the subclasses as necessary.
+     */
+    protected void listen() {
+        // nothing to do, intentionally empty
```

Comment:
@simonjenga this method couold be abstract if you would just inherit from `BaseFacade`